types/src/v0/impls/stake_table.rs

```diff
@@ -974,6 +981,136 @@ impl StakeTableFetcher {
         // Process the sorted events and return the resulting stake table.
         validators_from_l1_events(sorted.into_iter().map(|(_, e)| e))
     }
+    /// This function is used to calculate the reward for a block.
+    /// It fetches the initial supply from the token contract.
+    ///
+    /// Assumptions:
+    /// - This assumes only one mint event** occurred before the stake table was deployed.
+    /// - This is safe for now, because further minting is disabled until rewards withdrawal is implemented.
+    /// - However, once we support reward withdrawals, new mints will likely also come from the zero address**,
+    ///   making this logic unsafe.
+    ///
+    /// Safer approach:
+    /// - We now ely on the `Initialized` event of the token contract (which should only occur once).
+    /// - After locating this event, we **fetch its transaction receipt and look for a decoded `Transfer` log** (mint).
+    /// - If either step fails, the function aborts to prevent incorrect reward calculations.
+    pub async fn fetch_block_reward(&self) -> anyhow::Result<RewardAmount> {
+        let max_events_range = self.l1_client.options().l1_events_max_block_range;
+        let chain_config = *self.chain_config.lock().await;
+
+        let Some(stake_table_contract) = chain_config.stake_table_contract else {
+            bail!("No stake table contract address found in Chain config");
+        };
+
+        let provider = self.l1_client.provider.clone();
+        let stake_table = StakeTableV2::new(stake_table_contract, provider.clone());
+
+        let stake_table_init_block = stake_table
+            .initializedAtBlock()
+            .block(BlockId::finalized())
+            .call()
+            .await?
+            ._0
+            .to::<u64>();
+
+        tracing::info!("stake table init block ={stake_table_init_block}");
+
+        let token_address = stake_table
+            .token()
+            .block(BlockId::finalized())
+            .call()
+            .await
+            .context("Failed to get token address")?
+            ._0;
+
+        let token = EspToken::new(token_address, provider.clone());
+
+        // Attempt to find Initialized event
+        let init_logs = token
+            .Initialized_filter()
+            .from_block(0u64)
+            .to_block(BlockNumberOrTag::Finalized)
```

Comment:
Won't this fail due to rate limiting on some providers. I don't think we will get `None` in that case, I think we get an error, don't we?